File: tests_zig/linker.zig
const sd = @import("static_dynamic");
const utils = @import("util.zig");

pub const _start = {};

export fn main(argc: u64, argv: [*]const [*:0]const u8) callconv(.c) i32 {
    sd.init(argc, argv);

    utils.assert(sd.got.result.success != 0);
    utils.assert(@as(u64, @intFromPtr(sd.got.fns.dlopen))  != 0);
    utils.assert(@as(u64, @intFromPtr(sd.got.fns.dlsym))   != 0);
    utils.assert(@as(u64, @intFromPtr(sd.got.fns.dlclose)) != 0);
    utils.assert(@as(u64, @intFromPtr(sd.got.fns.dlerror)) != 0);

    const libc = sd.got.fns.dlopen("libc.so.6", .{ .NOW = true });
    utils.assert(libc != null);
    utils.assert(sd.got.fns.dlsym(libc, "printf") != null);

    utils.assert(sd.got.fns.dlsym(libc, "blah") == null);
    const err = sd.got.fns.dlerror();
    utils.assert(err != null and err[0] != 0);
    utils.assert(sd.got.fns.dlerror() == null);

    utils.assert(sd.got.fns.dlopen("blah.so", .{ .NOW = true }) == null);
    utils.assert(sd.got.fns.dlerror() != null);

    return 0;
}

File: tests_c/linker.c
#include "util.h"

int main(int argc, char** argv) {
  (void)argc;
  (void)argv;

  ASSERT(sd_got.success != 0);
  ASSERT(sd_got.dlopen != 0);
  ASSERT(sd_got.dlsym != 0);
  ASSERT(sd_got.dlclose != 0);
  ASSERT(sd_got.dlerror != 0);

  void* libc = sd_got.dlopen("libc.so.6", SD_RTLD_NOW);
  ASSERT(libc != 0);
  ASSERT(sd_got.dlsym(libc, "printf") != 0);

  ASSERT(sd_got.dlsym(libc, "blah") == 0);
  char* err = sd_got.dlerror();
  ASSERT(err != 0);
  ASSERT(sd_got.dlerror() == 0);

  ASSERT(sd_got.dlopen("blah.so", SD_RTLD_NOW) == 0);
  ASSERT(sd_got.dlerror() != 0);

  return 0;
}
